Add interaction policy and quoting controls to ActivityPub

pfefferle opened the pull request:
Introduces support for interaction policy fields in ActivityPub objects, updates the JSON-LD context to include GotoSocial and Mastodon extensions, and adds a 'Who can quote' select control to the editor plugin. Updates tests and ensures the new fields are handled in activity and base object classes.

<img width="391" height="266" alt="Screenshot 2025-09-18 at 15 49 14" src="https://github.com/user-attachments/assets/db003b37-2295-4e9c-a21f-248cd974fab7" />

Related: #181

See:

* https://docs.gotosocial.org/en/latest/federation/interaction_policy/
* https://blog.joinmastodon.org/2025/09/introducing-quote-posts/
* https://wordpress.org/support/topic/feature-request-support-for-mastodon-quotes/

## Proposed changes:
- Adds interaction policy metadata registration and validation
- Implements quote policy controls in the WordPress editor
- Updates JSON-LD context with GotoSocial extensions for interaction policies

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to the editor
* Change the "Who can quote" settings
* Check the JSON output for the corresponding JSON extension 

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [X] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Added a setting to control who can quote your posts.

</details>

pfefferle committed:
Add interaction policy and quoting controls to ActivityPub

Introduces support for interaction policy fields in ActivityPub objects, updates the JSON-LD context to include GotoSocial and Mastodon extensions, and adds a 'Who can quote' select control to the editor plugin. Updates tests and ensures the new fields are handled in activity and base object classes.

pfefferle committed:
Add interaction policy for quoting posts

Introduces new constants and post meta for 'activitypub_interaction_policy_quote' to control who can quote a post in the fediverse. Updates backend to register and sanitize the new meta, and updates the editor plugin UI to allow users to select the quoting policy.

pfefferle committed:
Add interaction policy controls to editor plugin

Introduces a new 'Interaction Policy' section in the editor plugin, allowing users to control who can quote their post on the fediverse. Updates the help text for the quote option and ensures the selected value is stored in post meta.

pfefferle committed:
Add get_interaction_policy method to Post class

Introduces a new get_interaction_policy() method to the Post class, allowing retrieval of the post's interaction policy based on actor mode and post meta. This supports more granular control over who can interact with posts, such as quoting, in line with federation standards.

pfefferle committed:
Update interaction policy help text in editor plugin

Revises the help text for the 'Interaction Policy' and 'Who can quote' controls in the ActivityPub editor plugin for improved clarity. The new text is more concise and user-focused.

pfefferle committed:
Merge branch 'trunk' into add/prepare-interaction-policy-support

pfefferle committed:
Merge branch 'trunk' into add/prepare-interaction-policy-support

pfefferle committed:
Improve interaction policy handling and add tests

Refactored the switch statement in Post::get_interaction_policy() to handle empty and invalid values by falling back to the 'anyone' policy. Added comprehensive unit tests to verify correct policy mapping for 'anyone', 'followers', 'me', missing, and invalid permission values.

matticbot committed:
Add changelog

Jiwoon-Kim commented:
https://github.com/Automattic/wordpress-activitypub/issues/1888

obenland commented on includes/transformer/class-post.php:
```diff
@@ -74,6 +74,68 @@ public function get_content_visibility() {
+	 *
+	 * @see https://docs.gotosocial.org/en/latest/federation/interaction_policy/
+	 *
+	 * @return array The interaction policy.
```
```suggestion
	 * @return array|null The interaction policy.
```

pfefferle replied:
technically every attribute can be `null`, so we would have to add it to every doc!?

Context: `null` is removing it from the output.

obenland commented on src/editor-plugin/plugin.js:
```diff
@@ -150,6 +158,26 @@ const EditorPlugin = () => {
 				className="activitypub-visibility"
 			/>
+
+			<BaseControl
+				label={ __( 'Interaction Policy', 'activitypub' ) }
+				help={ __( 'Choose who can interact with this post.', 'activitypub' ) }
+			></BaseControl>
+
+			<SelectControl
+				label={ __( 'Who can quote', 'activitypub' ) }
```
```suggestion
			<SelectControl
				label={ __( 'Who can quote this post?', 'activitypub' ) }
				help={ __( 'Quoting allows others to reshare your post while adding their own commentary.', 'activitypub' ) }
```

The two controls attempting to create hierarchy feel a bit awkward. Could we try playing with label and help in the select control to create the necessary context for this to work?

pfefferle replied:
I am also not very happy with it. I tried a lot of different versions, but hoped this one would be the best compromise.

The idea behind that was, that there might be other "Interaction Policy" controls coming, so it might be nice to somehow group the.

pfefferle committed:
Update property types to allow null in activity classes

Expanded property type annotations in Activity, Actor, Base_Object, and Generic_Object classes to include null where appropriate. Updated docblocks for getter and setter methods to reflect new types, improving type safety and clarity for optional properties.

pfefferle committed:
Update src/editor-plugin/plugin.js

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Merge branch 'trunk' into add/prepare-interaction-policy-support

matticbot committed:
Prettify

pfefferle committed:
Format help text for SelectControl in editor plugin

Refactored the help prop for the 'Who can quote this post?' SelectControl to use a multi-line format for improved readability and maintainability.

pfefferle committed:
Add Interaction Policy link to federation docs

Included a reference to the Interaction Policy documentation in the list of followed specifications to provide more comprehensive federation guidance.

obenland commented on includes/constants.php:
```diff
@@ -74,6 +74,10 @@
+// Interaction policy constants.
+define( 'ACTIVITYPUB_INTERACTION_POLICY_ANYONE', 'anyone' );
+define( 'ACTIVITYPUB_INTERACTION_POLICY_FOLLOWERS', 'followers' );
+define( 'ACTIVITYPUB_INTERACTION_POLICY_ME', 'me' );
```
If we expect other interaction policy items, should we make these quote-specific?

pfefferle replied:
My idea was to be able to re-use them for every policy! They post_meta is what makes it specific.

obenland replied:
Gotcha, I didn't realize these were applicable to multiple policies

obenland commented on includes/class-blocks.php:
```diff
@@ -83,6 +83,29 @@ public static function register_postmeta() {
+						}
+
+						return $value;
+					},
```
No action needed, just thinking out loud:

For sanitize callbacks like that, shouldn't we be able to define a schema in `show_in_rest` and it does all of that for us? Does it not fallback to the default if the new value doesn't fit the enum? There was _something_ that didn't work, but I don't remember what it was.

pfefferle replied:
I have no idea if that works, but happy to test it in a follow up!

obenland committed:
Docs updates

pfefferle committed:
Merge branch 'trunk' into add/prepare-interaction-policy-support

obenland committed:
Refactor interaction policy logic for quoting (#2214)

Jiwoon-Kim commented:
https://github.com/misskey-dev/misskey/issues/16590
